In [6]:
import numpy as np
import pandas as pd


In [7]:
books = pd.read_csv('Books.csv')
users = pd.read_csv('Users.csv')
ratings = pd.read_csv('Ratings.csv')

C:\Users\HP\AppData\Local\Temp\ipykernel_5824\1748781244.py:1: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  books = pd.read_csv('Books.csv')


In [8]:
books.head()


ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [9]:
users.head()


User-ID                            Location   Age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN

In [10]:
ratings.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [11]:
print(books.shape)
print(ratings.shape)
print(users.shape)

(271360, 8)
(1149780, 3)
(278858, 3)


In [12]:
books.isnull().sum()

ISBN                   0
Book-Title             0
Book-Author            1
Year-Of-Publication    0
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            3
dtype: int64

In [13]:
users.isnull().sum()

User-ID          0
Location         0
Age         110762
dtype: int64

In [14]:
ratings.isnull().sum()

User-ID        0
ISBN           0
Book-Rating    0
dtype: int64

In [15]:
print(books.duplicated().sum(),
users.duplicated().sum(),
ratings.duplicated().sum())

0 0 0


# Popularity Based Recommender System

In [16]:
ratings_with_name = ratings.merge(books,on='ISBN')

In [17]:
num_rating_df = ratings_with_name.groupby('Book-Title').count()['Book-Rating'].reset_index()
num_rating_df.rename(columns={'Book-Rating':'num_ratings'},inplace=True)
num_rating_df.head()

Book-Title  num_ratings
0   A Light in the Storm: The Civil War Diary of ...            4
1                              Always Have Popsicles            1
2               Apple Magic (The Collector's series)            1
3   Ask Lily (Young Women of Faith: Lily Series, ...            1
4   Beyond IBM: Leadership Marketing and Finance ...            1

Now, we will replace the `num_ratings` by `avg_rating`
> average rating/sum of average of all

In [18]:
avg_rating_df = ratings_with_name.groupby('Book-Title').mean()['Book-Rating'].reset_index()
avg_rating_df.rename(columns={'Book-Rating':'avg_rating'},inplace=True)
avg_rating_df.head()

Book-Title  avg_rating
0   A Light in the Storm: The Civil War Diary of ...        2.25
1                              Always Have Popsicles        0.00
2               Apple Magic (The Collector's series)        0.00
3   Ask Lily (Young Women of Faith: Lily Series, ...        8.00
4   Beyond IBM: Leadership Marketing and Finance ...        0.00

Now, We will be combining both columns `avg_rating`, `num-ratings`
on `Book-Title`

In [19]:
popular_df = num_rating_df.merge(avg_rating_df, on='Book-Title')
popular_df.head()

Book-Title  num_ratings  avg_rating
0   A Light in the Storm: The Civil War Diary of ...            4        2.25
1                              Always Have Popsicles            1        0.00
2               Apple Magic (The Collector's series)            1        0.00
3   Ask Lily (Young Women of Faith: Lily Series, ...            1        8.00
4   Beyond IBM: Leadership Marketing and Finance ...            1        0.00

Now, we will keep only those books whose rating is above 250

In [20]:
popular_df[popular_df['num_ratings']>=250]

Book-Title  num_ratings  \
764                                                  1984          284   
818                                   1st to Die: A Novel          509   
1048                                           2nd Chance          356   
1760                                   A Bend in the Road          346   
2281    A Child Called \It\": One Child's Courage to S...          265   
...                                                   ...          ...   
233850                           White Oleander : A Novel          387   
233851       White Oleander : A Novel (Oprah's Book Club)          356   
234740  Wicked: The Life and Times of the Wicked Witch...          326   
234951                                        Wild Animus         2502   
240860                                \O\" Is for Outlaw"          269   

        avg_rating  
764       4.454225  
818       3.575639  
1048      3.269663  
1760      3.364162  
2281      4.086792  
...            ...  
233850    3.506460  
233851    3.772472  
234740    3.766871  
234951    1.019584  
240860    2.988848  

[186 rows x 3 columns]

In [21]:
popular_df = popular_df[popular_df['num_ratings']>=250].sort_values('avg_rating', ascending=False).head(50)
popular_df.head()

Book-Title  num_ratings  \
80434  Harry Potter and the Prisoner of Azkaban (Book 3)          428   
80422       Harry Potter and the Goblet of Fire (Book 4)          387   
80441     Harry Potter and the Sorcerer's Stone (Book 1)          278   
80426  Harry Potter and the Order of the Phoenix (Boo...          347   
80414   Harry Potter and the Chamber of Secrets (Book 2)          556   

       avg_rating  
80434    5.852804  
80422    5.824289  
80441    5.737410  
80426    5.501441  
80414    5.183453

In [22]:
popular_df.merge(books, on='Book-Title')

Book-Title  num_ratings  \
0    Harry Potter and the Prisoner of Azkaban (Book 3)          428   
1    Harry Potter and the Prisoner of Azkaban (Book 3)          428   
2    Harry Potter and the Prisoner of Azkaban (Book 3)          428   
3         Harry Potter and the Goblet of Fire (Book 4)          387   
4         Harry Potter and the Goblet of Fire (Book 4)          387   
..                                                 ...          ...   
191                                  Lord of the Flies          259   
192                            A Prayer for Owen Meany          413   
193                            A Prayer for Owen Meany          413   
194   The Vampire Lestat (Vampire Chronicles, Book II)          301   
195       White Oleander : A Novel (Oprah's Book Club)          356   

     avg_rating        ISBN         Book-Author Year-Of-Publication  \
0      5.852804  0439136350       J. K. Rowling                1999   
1      5.852804  0439136369       J. K. Rowling                2001   
2      5.852804  0786222743       J. K. Rowling                2000   
3      5.824289  0439139597       J. K. Rowling                2000   
4      5.824289  0439139600       J. K. Rowling                2002   
..          ...         ...                 ...                 ...   
191    3.818533  0698102193  William G. Golding                1978   
192    3.796610  0345361792         John Irving                1990   
193    3.796610  0345363523         John Irving                1989   
194    3.777409  0345313860           ANNE RICE                1986   
195    3.772472  0316284955         Janet Fitch                2000   

                 Publisher                                        Image-URL-S  \
0               Scholastic  http://images.amazon.com/images/P/0439136350.0...   
1               Scholastic  http://images.amazon.com/images/P/0439136369.0...   
2          Thorndike Press  http://images.amazon.com/images/P/0786222743.0...   
3               Scholastic  http://images.amazon.com/images/P/0439139597.0...   
4    Scholastic Paperbacks  http://images.amazon.com/images/P/0439139600.0...   
..                     ...                                                ...   
191       Putnam Pub Group  http://images.amazon.com/images/P/0698102193.0...   
192       Ballantine Books  http://images.amazon.com/images/P/0345361792.0...   
193       Ballantine Books  http://images.amazon.com/images/P/0345363523.0...   
194       Ballantine Books  http://images.amazon.com/images/P/0345313860.0...   
195         Back Bay Books  http://images.amazon.com/images/P/0316284955.0...   

                                           Image-URL-M  \
0    http://images.amazon.com/images/P/0439136350.0...   
1    http://images.amazon.com/images/P/0439136369.0...   
2    http://images.amazon.com/images/P/0786222743.0...   
3    http://images.amazon.com/images/P/0439139597.0...   
4    http://images.amazon.com/images/P/0439139600.0...   
..                                                 ...   
191  http://images.amazon.com/images/P/0698102193.0...   
192  http://images.amazon.com/images/P/0345361792.0...   
193  http://images.amazon.com/images/P/0345363523.0...   
194  http://images.amazon.com/images/P/0345313860.0...   
195  http://images.amazon.com/images/P/0316284955.0...   

                                           Image-URL-L  
0    http://images.amazon.com/images/P/0439136350.0...  
1    http://images.amazon.com/images/P/0439136369.0...  
2    http://images.amazon.com/images/P/0786222743.0...  
3    http://images.amazon.com/images/P/0439139597.0...  
4    http://images.amazon.com/images/P/0439139600.0...  
..                                                 ...  
191  http://images.amazon.com/images/P/0698102193.0...  
192  http://images.amazon.com/images/P/0345361792.0...  
193  http://images.amazon.com/images/P/0345363523.0...  
194  http://images.amazon.com/images/P/0345313860.0...  
195  http://images.amazon.com/images/P/03

In [23]:
popular_df.merge(books, on='Book-Title').shape

(196, 10)

196 because, there are various books on Harry Potter, but they have differend ISB number

In [85]:
popular_df = popular_df.merge(books, on='Book-Title').drop_duplicates('Book-Title')[['Book-Title', 'Book-Author', 'Image-URL-M', 'num_ratings', 'avg_rating']]

This is the list of top 50 books to be displayed on website


# Collaborative Filtering Based Recommender System

Now, we will select only those users who have rated 200+ books 
> To select only active and experienced readers

In [25]:
ratings_with_name.groupby('User-ID').count()['Book-Rating'] > 200

User-ID
2         False
8         False
9         False
10        False
12        False
          ...  
278846    False
278849    False
278851    False
278852    False
278854    False
Name: Book-Rating, Length: 92106, dtype: bool

In [100]:
#to print this in boolean and print only those who have read above 200
x = ratings_with_name.groupby('User-ID').count()['Book-Rating'] > 200
experienced_users = x[x].index

We will now fetch only `experienced_users` form `User-ID` 

In [101]:
filtered_rating = ratings_with_name[ratings_with_name['User-ID'].isin(experienced_users)]


Now that we have completed the user based filtering,, we will now do book based filtering 

In [102]:
filtered_rating.head()

User-ID        ISBN  Book-Rating            Book-Title Book-Author  \
2      6543  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
5     23768  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
7     28523  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
15    77940  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
16    81977  034545104X            0  Flesh Tones: A Novel  M. J. Rose   

   Year-Of-Publication         Publisher  \
2                 2002  Ballantine Books   
5                 2002  Ballantine Books   
7                 2002  Ballantine Books   
15                2002  Ballantine Books   
16                2002  Ballantine Books   

                                          Image-URL-S  \
2   http://images.amazon.com/images/P/034545104X.0...   
5   http://images.amazon.com/images/P/034545104X.0...   
7   http://images.amazon.com/images/P/034545104X.0...   
15  http://images.amazon.com/images/P/034545104X.0...   
16  http://images.amazon.com/images/P/034545104X.0...   

                                          Image-URL-M  \
2   http://images.amazon.com/images/P/034545104X.0...   
5   http://images.amazon.com/images/P/034545104X.0...   
7   http://images.amazon.com/images/P/034545104X.0...   
15  http://images.amazon.com/images/P/034545104X.0...   
16  http://images.amazon.com/images/P/034545104X.0...   

                                          Image-URL-L  
2   http://images.amazon.com/images/P/034545104X.0...  
5   http://images.amazon.com/images/P/034545104X.0...  
7   http://images.amazon.com/images/P/034545104X.0...  
15  http://images.amazon.com/images/P/034545104X.0...  
16  http://images.amazon.com/images/P/034545104X.0...

we will do it on the basis of book title

In [103]:
y = filtered_rating.groupby('Book-Title').count()['Book-Rating']>=50
famous_books = y[y].index

In [104]:
final_rating = filtered_rating[filtered_rating['Book-Title'].isin(famous_books)]

In [105]:
final_rating.head()

User-ID        ISBN  Book-Rating    Book-Title      Book-Author  \
63   278418  0446520802            0  The Notebook  Nicholas Sparks   
65     3363  0446520802            0  The Notebook  Nicholas Sparks   
66     7158  0446520802           10  The Notebook  Nicholas Sparks   
69    11676  0446520802           10  The Notebook  Nicholas Sparks   
74    23768  0446520802            6  The Notebook  Nicholas Sparks   

   Year-Of-Publication     Publisher  \
63                1996  Warner Books   
65                1996  Warner Books   
66                1996  Warner Books   
69                1996  Warner Books   
74                1996  Warner Books   

                                          Image-URL-S  \
63  http://images.amazon.com/images/P/0446520802.0...   
65  http://images.amazon.com/images/P/0446520802.0...   
66  http://images.amazon.com/images/P/0446520802.0...   
69  http://images.amazon.com/images/P/0446520802.0...   
74  http://images.amazon.com/images/P/0446520802.0...   

                                          Image-URL-M  \
63  http://images.amazon.com/images/P/0446520802.0...   
65  http://images.amazon.com/images/P/0446520802.0...   
66  http://images.amazon.com/images/P/0446520802.0...   
69  http://images.amazon.com/images/P/0446520802.0...   
74  http://images.amazon.com/images/P/0446520802.0...   

                                          Image-URL-L  
63  http://images.amazon.com/images/P/0446520802.0...  
65  http://images.amazon.com/images/P/0446520802.0...  
66  http://images.amazon.com/images/P/0446520802.0...  
69  http://images.amazon.com/images/P/0446520802.0...  
74  http://images.amazon.com/images/P/0446520802.0...

There are no duplicates on the dataset, you may check

In [106]:
pt = final_rating.pivot_table(index='Book-Title', columns='User-ID', values='Book-Rating')

In [107]:
pt.head()

User-ID              254     2276    2766    2977    3363    4017    4385    \
Book-Title                                                                    
1984                    9.0     NaN     NaN     NaN     NaN     NaN     NaN   
1st to Die: A Novel     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
2nd Chance              NaN    10.0     NaN     NaN     NaN     NaN     NaN   
4 Blondes               NaN     NaN     NaN     NaN     NaN     NaN     NaN   
A Bend in the Road      0.0     NaN     7.0     NaN     NaN     NaN     NaN   

User-ID              6251    6323    6543    ...  271705  273979  274004  \
Book-Title                                   ...                           
1984                    NaN     NaN     NaN  ...    10.0     NaN     NaN   
1st to Die: A Novel     NaN     NaN     9.0  ...     NaN     NaN     NaN   
2nd Chance              NaN     NaN     0.0  ...     NaN     NaN     NaN   
4 Blondes               0.0     NaN     NaN  ...     NaN     NaN     NaN   
A Bend in the Road      NaN     NaN     NaN  ...     NaN     0.0     NaN   

User-ID              274061  274301  274308  275970  277427  277639  278418  
Book-Title                                                                   
1984                    NaN     NaN     NaN     0.0     NaN     NaN     NaN  
1st to Die: A Novel     NaN     NaN     NaN     NaN     NaN     NaN     NaN  
2nd Chance              NaN     NaN     0.0     NaN     NaN     0.0     NaN  
4 Blondes               NaN     NaN     NaN     NaN     NaN     NaN     NaN  
A Bend in the Road      NaN     NaN     NaN     NaN     NaN     NaN     NaN  

[5 rows x 810 columns]

In [108]:
pt.fillna(0, inplace=True)

In [109]:
pt.head()

User-ID              254     2276    2766    2977    3363    4017    4385    \
Book-Title                                                                    
1984                    9.0     0.0     0.0     0.0     0.0     0.0     0.0   
1st to Die: A Novel     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
2nd Chance              0.0    10.0     0.0     0.0     0.0     0.0     0.0   
4 Blondes               0.0     0.0     0.0     0.0     0.0     0.0     0.0   
A Bend in the Road      0.0     0.0     7.0     0.0     0.0     0.0     0.0   

User-ID              6251    6323    6543    ...  271705  273979  274004  \
Book-Title                                   ...                           
1984                    0.0     0.0     0.0  ...    10.0     0.0     0.0   
1st to Die: A Novel     0.0     0.0     9.0  ...     0.0     0.0     0.0   
2nd Chance              0.0     0.0     0.0  ...     0.0     0.0     0.0   
4 Blondes               0.0     0.0     0.0  ...     0.0     0.0     0.0   
A Bend in the Road      0.0     0.0     0.0  ...     0.0     0.0     0.0   

User-ID              274061  274301  274308  275970  277427  277639  278418  
Book-Title                                                                   
1984                    0.0     0.0     0.0     0.0     0.0     0.0     0.0  
1st to Die: A Novel     0.0     0.0     0.0     0.0     0.0     0.0     0.0  
2nd Chance              0.0     0.0     0.0     0.0     0.0     0.0     0.0  
4 Blondes               0.0     0.0     0.0     0.0     0.0     0.0     0.0  
A Bend in the Road      0.0     0.0     0.0     0.0     0.0     0.0     0.0  

[5 rows x 810 columns]

## Preprocessing of the data has now been done

### Now, we will make a model

`Each book is a point(vector) in 810 dimension coordinate system`

All you have to do is to find the distances from one to every other vector
> Eucledian Distance 

we will now import cosine similarity metric from sklearn

In [36]:
!pip install scikit-learn 

Was unable to get the library, so had to create the enviroment folder first

In [110]:
import sklearn

In [111]:
from sklearn.metrics.pairwise import cosine_similarity

In [112]:
cosine_similarity(pt).shape

(706, 706)

`(706,706)` signifies that we have got the Eucledean distance between 706 and 706 vectors 

In [113]:
similarity_scores = cosine_similarity(pt)
similarity_scores

array([[1.        , 0.10255025, 0.01220856, ..., 0.12110367, 0.07347567,
        0.04316046],
       [0.10255025, 1.        , 0.2364573 , ..., 0.07446129, 0.16773875,
        0.14263397],
       [0.01220856, 0.2364573 , 1.        , ..., 0.04558758, 0.04938579,
        0.10796119],
       ...,
       [0.12110367, 0.07446129, 0.04558758, ..., 1.        , 0.07085128,
        0.0196177 ],
       [0.07347567, 0.16773875, 0.04938579, ..., 0.07085128, 1.        ,
        0.10602962],
       [0.04316046, 0.14263397, 0.10796119, ..., 0.0196177 , 0.10602962,
        1.        ]])

Found the distance between every book and every other book

Now, we will make a function which will take a book name as an input and give 5 other books (recommendations) as output

In [114]:
np.where(pt.index=='1984')[0][0]

0

In [115]:
np.where(pt.index=='2nd Chance')[0][0]

2

In [116]:
def recommend(book_name):
    
    # Job-1 : Fetch the Index
    index = np.where(pt.index==book_name)[0][0]
    
    # Job-2 : Find out the distance
    distance = similarity_scores[index]
    
    # Job-3 : Find the similar items
    similar_items = sorted(list(enumerate(similarity_scores[index])), key=lambda x:x[1], reverse=True)[1:6]
    
    data = []
    for i in similar_items:
        item = []
        temp_df = books[books['Book-Title'] == pt.index[i[0]]]
        item.extend(list(temp_df.drop_duplicates('Book-Title')['Book-Title'].values))
        item.extend(list(temp_df.drop_duplicates('Book-Title')['Book-Author'].values))
        item.extend(list(temp_df.drop_duplicates('Book-Title')['Image-URL-M'].values))
        
        data.append(item)
        return data

In [117]:
sorted(list(enumerate(similarity_scores[0])))
# The similarity (distance) of all other books with the one in index 0

[(0, 0.9999999999999994),
 (1, 0.10255024829874465),
 (2, 0.012208555646509366),
 (3, 0.0),
 (4, 0.053672244225442485),
 (5, 0.027749013024662858),
 (6, 0.08216491130000354),
 (7, 0.13732869060263378),
 (8, 0.03261686350494277),
 (9, 0.03667590746512177),
 (10, 0.02322418425912186),
 (11, 0.06766486753662343),
 (12, 0.020839782065795794),
 (13, 0.09673734976610586),
 (14, 0.13388864701187034),
 (15, 0.08303111691811028),
 (16, 0.1115354275858341),
 (17, 0.051004105100675136),
 (18, 0.025177841632809915),
 (19, 0.11706383185253948),
 (20, 0.0),
 (21, 0.14333792815336935),
 (22, 0.07847534327232127),
 (23, 0.06150451339525138),
 (24, 0.08723968326958671),
 (25, 0.0),
 (26, 0.07009814362931202),
 (27, 0.1365868085547294),
 (28, 0.07600327535501016),
 (29, 0.12167133598776537),
 (30, 0.007680459111722776),
 (31, 0.01473221255132234),
 (32, 0.0),
 (33, 0.07965814289647455),
 (34, 0.04522616573040869),
 (35, 0.01556270914112474),
 (36, 0.09495937740802214),
 (37, 0.018230696811058135),
 (38,

* enumerate: To give the numberto the list
* Also, Note that the above sorting is on the basis of index, we will have to introduce lambda 

In [118]:
sorted(list(enumerate(similarity_scores[0])), key=lambda x:x[1], reverse=True)[1:6]

[(47, 0.2702651417103732),
 (545, 0.26396193711234966),
 (82, 0.2366937434740099),
 (634, 0.23299389358170394),
 (551, 0.2262639743141286)]

In [119]:
recommend('1984')

[['Animal Farm',
  'George Orwell',
  'http://images.amazon.com/images/P/0451526341.01.MZZZZZZZ.jpg']]

# End of the Model, Now go to vscode and create the webpage

In [120]:
final_rating['Image-URL-M'][63]

'http://images.amazon.com/images/P/0446520802.01.MZZZZZZZ.jpg'

# Now, we will use pickle to export 

In [121]:
import pickle
pickle.dump(popular_df, open('popular.pkl', 'wb'))

In [122]:
pickle.dump(pt, open('pt.pkl','wb'))
pickle.dump(books, open('books.pkl','wb'))
pickle.dump(similarity_scores, open('similarity-scores.pkl', 'wb'))

In [123]:
popular_df.head()

Book-Title    Book-Author  \
0   Harry Potter and the Prisoner of Azkaban (Book 3)  J. K. Rowling   
3        Harry Potter and the Goblet of Fire (Book 4)  J. K. Rowling   
5      Harry Potter and the Sorcerer's Stone (Book 1)  J. K. Rowling   
9   Harry Potter and the Order of the Phoenix (Boo...  J. K. Rowling   
13   Harry Potter and the Chamber of Secrets (Book 2)  J. K. Rowling   

                                          Image-URL-M  num_ratings  avg_rating  
0   http://images.amazon.com/images/P/0439136350.0...          428    5.852804  
3   http://images.amazon.com/images/P/0439139597.0...          387    5.824289  
5   http://images.amazon.com/images/P/0590353403.0...          278    5.737410  
9   http://images.amazon.com/images/P/043935806X.0...          347    5.501441  
13  http://images.amazon.com/images/P/0439064872.0...          556    5.183453